tests/functional/login.js

```diff
@@ -0,0 +1,80 @@
+const Nightmare = require('nightmare');
+const assert = require('chai').assert;
+
+const url = 'http://localhost:3333/';
+
+
+
+describe('Login tests', function() {
+
+  this.timeout('30s');
+
+  let browser = null;
+
+  beforeEach( () => {
+    browser = new Nightmare({ 
+      // to make the typed input much faster
+      typeInterval: 20,
+      // uncomment this to show the browser and the debug window
+      // openDevTools: {
+      //   mode: "detach"
+      // },
+      // show: true 
+    });
+    browser
+      .goto(url);
+  });
+
+  describe('Functional tests for login', () => {
+  
+    it('we should be redirected to the login page', done => {
+      browser
+        .end()
+        .evaluate( () => { return document; } )
+        .then(function (document) { 
+          assert.equal(document.location.href, 'http://localhost:3333/login');
+          done();
+        })
+        .catch(done);
+    });
+
+    it('we cannot login with false credentials', done => {
+      var selector = '.notice-wrapper';
+      browser
+        .type('#username', 'sald')
+        .type('#password', 'sald')
+        .click('#login-submit')
+        .wait('.notice')
+        .end()
+        .evaluate(selector => {
+          return document.querySelector('.notice-wrapper div').textContent;
+        }, selector)
+        .then(function (message) {
+          assert.equal(message, 'Authentication failed');
```

Comment:
It's a bit of a shame to have the tests rely on particular text which might change in the future, but I can also see the value in checking the error. Just thought I would say this to spark a discussion.

Reply:
I don't think it's a shame, it's actually really good. There are a many reasons why this is perfectly fine:

1) This test is incredibly easy to read as you immediately understand what is going on.
2) We expect this error message to be shown when you have the wrong credentials, so this is a logical thing to test for.
3) When the message is changed, or when the login sequence has changed (or when you made a mistake) this test will fail and forces you to investigate or to adjust the test.